hedera-node/docs/design/services/token-service/airdrops/airdrop-pending-state.md

```diff
@@ -0,0 +1,122 @@
+
+## Prerequisite reading
+* [HIP-904](https://hips.hedera.com/hip/hip-904)
+
+## Goals
+
+1. Define state structure where to keep the information for incoming airdrops, so that they could be canceled or claimed
+2. Define what state transitions inside the pending state would be allowed in the scope of airdrop operations and how changes would be processed
+3. Describe how this new state would be serialized and kept after restarts
+4. Define Readable and Writable stores for accessing and manipulating the pending airdrops state
+
+## Architecture
+
+The newly defined entity state won’t be able to persist anything in the state directly. If we want to commit something from the pending state to the persisted state, we would have to delegate a given pending airdrop by creating a synthetic crypto transfer transaction, send it to a suitable handler and delete it from the pending state. The writable nature of the pending state would be limited only to adding new entries for a pending airdrop to be transferred or to remove already added ones.
+
+The `WritableStore` of the airdrops, should contain a handler method and if a given pending airdrop is claimed, it could be delegated to the `CryptoTransferHandler` and executed.
+
+Details for `TokenClaimAirdrop` and `TokenCancelAirdrop` would be covered in their own design, docs but basically the `claim` operation would take a pending airdrop and submit it to the delegate function for actual handling and execution. The `cancel` operation would clear an added pending airdrop from the state and it won’t be executed at all.
+
+### Structure
+
+We should define the structure of pending changes that would be kept in the `InitialModServiceTokenSchema` state, so we would reuse the token’s serialized state and add pending airdrops records in it. We could use the type defined in the protobuf records for the airdrop information itself. We should use the protobuf types since they are used for state serialization. So, the structure could look like this:
+
+`StateDefinition.*onDisk*(*AIRDROPS_KEY*, PendingAirdropId.PROTOBUF, PendingAirdropValue.PROTOBUF, MAX__AIRDROPS);`
+
+We can have a case, where we have multiple airdrops for the same fungible token. In this case the  `PendingAirdropValue` would be updated with the new aggregated value.
+
+### Stores
+
+We need to define a special `WritableStore` for the airdrops, so that we can manipulate the content in them. This would be the `WritableAirdropStore`.
+
+It would extend `ReadableAirdropStoreImpl` that implements the following interface:
+
+```java
+public interface ReadableAirdropStore {
+	PendingAirdropValue get(@NonNull final PendingAirdropId airdropId);
```

Comment:
What would this method return in the case of an NFT?

Reply:
It won't be applicable for NFTs. I renamed the method, so that it's clear it will be used only for fungible tokens.